File: services/song_service.py
from models.song import Song
from services.audio_service import AudioService


class SongService:

    def __init__(self, callback):
        self._index = 0
        self._callback = callback
        self._songs = []
        self._audio_service = None

    def set_songs(self, songs: []):
        self._songs = songs
        self._index = 0
        if len(self._songs) > 0:
            self._init_audio()

    def _init_audio(self):
        song = self.current()
        if self._audio_service is not None:
            self._audio_service.stop()
        
        self._audio_service = AudioService(song, self._callback)

    def forward(self):
        if len(self._songs) == 0:
            return

        self._index = self._index+1
        self._index = 0 if self._index == len(
            self._songs) else self._index
        self._init_audio()

    def backward(self):
        if len(self._songs) == 0:
            return

        self._index = self._index-1
        self._index = len(self._songs)-1 if self._index < 0 else self._index
        self._init_audio()

    def current(self):
        if len(self._songs) == 0:
            return None
        return self._songs[self._index]

    def next_audio(self):
        if self._audio_service is None:
            return

        self._audio_service.forward()

    def current_audio(self):
        if self._audio_service is None:
            return None

        return self._audio_service.current()

    def play(self):
        if self._audio_service is None:
            return

        return self._audio_service.play()

    def stop(self):
        if self._audio_service is None:
            return

        return self._audio_service.stop()

    def forwardAudio(self):
        return self.next_audio()

    def currentAudio(self):
        return self.current_audio()


Song_Service = SongService

File: src/song_service.cpp
#include "song_service.hpp"

namespace cherry {

SongService::SongService(
    AudioPlayerFactory& player_factory,
    PlaybackSequencer& sequencer,
    PlayerService::EventCallback callback,
    LoadProgressCallback load_progress)
    : player_factory_(player_factory),
      sequencer_(sequencer),
      callback_(std::move(callback)),
      load_progress_(std::move(load_progress)) {}

void SongService::set_songs(std::vector<Song> songs) {
    songs_ = std::move(songs);
    index_ = 0;
    audio_service_.reset();

    if (!songs_.empty()) {
        init_audio(load_progress_);
    }
}

void SongService::init_audio(LoadProgressCallback progress) {
    if (audio_service_ != nullptr) {
        audio_service_->stop();
    }

    audio_service_ = std::make_unique<AudioService>(
        songs_[index_], player_factory_, sequencer_, callback_, std::move(progress));
}

void SongService::forward() {
    if (songs_.empty()) {
        return;
    }

    ++index_;
    if (index_ == songs_.size()) {
        index_ = 0;
    }

    init_audio();
}

void SongService::backward() {
    if (songs_.empty()) {
        return;
    }

    index_ = (index_ == 0) ? songs_.size() - 1 : index_ - 1;
    init_audio();
}

const Song* SongService::current() const {
    if (songs_.empty()) {
        return nullptr;
    }

    return &songs_[index_];
}

void SongService::next_audio() {
    if (audio_service_ != nullptr) {
        audio_service_->forward();
    }
}

const Audio* SongService::current_audio() const {
    if (audio_service_ == nullptr) {
        return nullptr;
    }

    return &audio_service_->current();
}

void SongService::play() {
    if (audio_service_ != nullptr) {
        audio_service_->play();
    }
}

void SongService::stop() {
    if (audio_service_ != nullptr) {
        audio_service_->stop();
    }
}

double SongService::current_progress() const {
    return audio_service_ != nullptr ? audio_service_->progress() : 0.0;
}

}  // namespace cherry
